src/pydicom/filebase.py

```diff
@@ -1,211 +1,332 @@
 
     @is_little_endian.setter
     def is_little_endian(self, value: bool) -> None:
+        """Get/set the endianness for encoding/decoding, ``True`` for little
+        endian and ``False`` for big endian.
+        """
+        if not isinstance(value, bool):
+            raise TypeError(f"'{type(self).__name__}.is_little_endian' must be bool")
+
         self._little_endian = value
-        if value:  # Little Endian
-            self.read_US = self.read_leUS
-            self.read_UL = self.read_leUL
-            self.write_US = self.write_leUS  # type: ignore[method-assign]
-            self.write_UL = self.write_leUL  # type: ignore[method-assign]
-            self.read_tag = self.read_le_tag
-        else:  # Big Endian
-            self.read_US = self.read_beUS
-            self.read_UL = self.read_beUL
-            self.write_US = self.write_beUS  # type: ignore[method-assign]
-            self.write_UL = self.write_beUL  # type: ignore[method-assign]
-            self.read_tag = self.read_be_tag
+
+        endianness = "><"[value]
+        self._us_packer = Struct(f"{endianness}H").pack
+        self._us_unpacker = Struct(f"{endianness}H").unpack
+        self._ul_packer = Struct(f"{endianness}L").pack
+        self._ul_unpacker = Struct(f"{endianness}L").unpack
+        self._tag_packer = Struct(f"{endianness}2H").pack
+        self._tag_unpacker = Struct(f"{endianness}2H").unpack
 
     @property
     def is_implicit_VR(self) -> bool:
-        return self._implicit_VR
+        if not hasattr(self, "_implicit_vr"):
+            raise AttributeError(
+                f"{type(self).__name__}.is_implicit_VR' has not been set"
+            )
+
+        return self._implicit_vr
 
     @is_implicit_VR.setter
     def is_implicit_VR(self, value: bool) -> None:
-        self._implicit_VR = value
+        """Get/set the VR mode for encoding/decoding. ``True`` for implicit VR
+        and ``False`` for explicit VR.
+        """
+        if not isinstance(value, bool):
+            raise TypeError(f"'{type(self).__name__}.is_implicit_VR' must be bool")
+
+        self._implicit_vr = value
+
+    @property
+    def name(self) -> str:
+        """Return the value of the :attr:`~pydicom.filebase.DicomIO.parent`'s
+        `name` attribute, or ``"<no filename>"`` if no such attribute.
+        """
+        return getattr(self.parent, "name", "<no filename>")
+
+    def read(self, size: int = -1) -> bytes:
+        """Read up to `size` bytes from the buffer and return them. If `size`
+        is unspecified, all bytes until EOF are returned.
+
+        Fewer than `size` bytes may be returned if the operating system call
+        returns fewer than `size` bytes.
+        """
+        name = type(self).__name__
+        buffer = type(self.parent).__name__
+        raise OSError(
+            f"'{name}' cannot be used with '{buffer}': object has no read() method"
+        )
+
+    def read_exact(self, length: int, nr_retries: int = 3) -> bytes:
```

Comment:
Most of the speed improvement in reading comes from splitting this out. It only gets used by `read_tag` and has the added bonus of making `dcmwrite(DicomBytesIO(), ds)` faster because  `DicomFileLike.parent_read` is now directly `BytesIO.read` rather than going `DicomIO.read -> BytesIO.read`